R/layers.R

```diff
@@ -30,6 +30,77 @@ layer <- function(subclass, ..., .prefix = "layer_") {
   structure(list(...), class = c(paste0(.prefix, subclass), "layer"))
 }
 
+#' Update post-processing `layer`
+#'
+#' This `layer` method for `update()` takes named arguments as `...` whose values
+#' will replace the elements of the same name in the actual post-processing layer.
+#' Analogous to `update.step()` from the `recipes` package.
+#'
+#' @param object A post-processing `layer`.
+#' @param ... Key-value pairs where the keys match up with names of elements
+#' in the layer, and the values are the new values to update the layer with.
+#'
+#' @examples
+#' jhu <- case_death_rate_subset %>%
+#' dplyr::filter(time_value > "2021-11-01", geo_value %in% c("ak", "ca", "ny"))
+#' r <- epi_recipe(jhu) %>%
+#'   step_epi_lag(death_rate, lag = c(0, 7, 14)) %>%
+#'   step_epi_ahead(death_rate, ahead = 7) %>%
+#'   step_epi_naomit()
+#' wf <- epi_workflow(r, parsnip::linear_reg()) %>% fit(jhu)
+#' latest <- jhu %>%
+#'   dplyr::filter(time_value >= max(time_value) - 14)
+#'
+#' # Specify a `forecast_date` that is greater than or equal to `as_of` date
+#' f <- frosting() %>% layer_predict() %>%
+#'   layer_add_forecast_date(forecast_date = "2022-05-31") %>%
+#'   layer_naomit(.pred)
+#'
+#' wf1 <- wf %>% add_frosting(f)
+#'
+#' p1 <- predict(wf1, latest)
+#' p1
+#'
+#' # Update forecast date
+#' f$layers[[2]] <- update(f$layers[[2]], forecast_date = "2021-06-01")
```

Comment:
I think we should discuss how this is implemented. I don't really like `recipes::update()` because you somehow have to know the order of the layers. I think it should instead look for a class. Perhaps we should have `update_layer()` and `update_epi_recipe()` as more nuanced wrappers around `update()`.

Reply:
Do you mean to have the user input the name of the layer (like in the `extract_argument` functions)? But then, what if there are two or more layers with the same name (similar to how there may be multiple lag steps)?

Reply:
Yes, that's what I mean. And I agree with the potential difficulty. But I feel like the existing version is far too annoying. I think we should brainstorm how to avoid possible bad behaviour while also making it easy for the user.

* There's also the case of trying to update a recipe in a workflow.

* An ideal test bed is something like:
  - take the fitted workflow from `arx_forecaster()`
  - add a recipe step in between other steps
  - change some of the existing steps and layers
  - refit the whole thing and produce predictions

See also #179. I think this is kind of a big project, and I'm not sure how to handle it yet. Maybe good to discuss in a future tooling meeting?

Reply:
This does sound better (and I think something like #179 would be doable), though I also think that it may be good to have something simple to use to be able to update the workflow with in the meantime (rather than the current approach of going down the rabbit hole of $ as in `axe$pre$actions$recipe$recipe$steps[[1]]` - see the discussion in #179)